tests/core/data/test_tensordense.py

```diff
@@ -0,0 +1,234 @@
+import numpy as np
+from numpy.testing import assert_almost_equal
+import tensorflow as tf
+import pytest
+
+from qutip.core import data
+from qutip.core.data import dense
+from qutip_tensorflow.core.data import DenseTensor
+
+from . import conftest
+
+
+# Set up some fixtures for automatic parametrisation.
+
+@pytest.fixture(params=[
+    pytest.param((1, 5), id='ket'),
+    pytest.param((5, 1), id='bra'),
+    pytest.param((5, 5), id='square'),
+    pytest.param((2, 4), id='wide'),
+    pytest.param((4, 2), id='tall'),
+])
+def shape(request): return request.param
+
+
+@pytest.fixture(params=[True, False], ids=['Fortran', 'C'])
+def fortran(request): return request.param
+
+
+def _valid_numpy():
+    # Arbitrary valid numpy array.
+    return conftest.random_numpy_dense((5, 5), False)
+
+def _valid_tensor():
+    # Arbitrary valid tensor array.
+    return conftest.random_tensor_dense((5, 5))
+
+def _valid_list():
+    # Arbitrary valid tensor array.
+    return conftest.random_numpy_dense((5, 5), False).tolist()
+
+
+@pytest.fixture(scope='function')
+def numpy_dense(shape, fortran):
+    return conftest.random_numpy_dense(shape, fortran)
+
+@pytest.fixture(scope='function')
+def tensor_dense(shape):
+    return conftest.random_tensor_dense(shape)
+
+
+@pytest.fixture(scope='function')
+def list_dense(shape):
+    return conftest.random_numpy_dense(shape, fortran).tolist()
+
+@pytest.fixture(scope='function')
+def data_tensor_dense(tensor_dense):
+    return DenseTensor(tensor_dense)
+
+
+class TestClassMethods:
+    def test_init_from_list(self, list_dense, shape):
+        test = DenseTensor(list_dense, shape)
+        assert test.shape == shape
+        assert np.all(test.to_array() == np.array(list_dense))
+
+    def test_init_from_ndarray(self, numpy_dense):
+        test = DenseTensor(numpy_dense)
+        assert test.shape == numpy_dense.shape
+        assert np.all(test.to_array() == numpy_dense)
+
+    def test_init_from_tensor(self, tensor_dense):
+        test = DenseTensor(tensor_dense)
+        assert test.shape == tuple(tensor_dense.shape.as_list())
+        assert np.all(test.to_array() == tensor_dense)
+
+        # by default we do not return a copy
+        assert test._tf is tensor_dense
+
+    @pytest.mark.parametrize('dtype', ['complex128',
+                                       'float64',
+                                       'int32', 'int64',
+                                       'uint32'])
+    def test_init_from_list_other_dtype(self, shape, dtype):
+        numpy_dense = np.random.rand(*shape).astype(dtype, casting='unsafe')
+        list_dense = numpy_dense.tolist()
+        test = DenseTensor(list_dense, numpy_dense.shape)
+        assert test.shape == shape
+        assert test._tf.dtype == tf.complex128
+        assert test._tf.shape == shape
+        assert_almost_equal(test.to_array(), list_dense)
```

Comment:
Are they really allowed to only be _almost_ equal?

Reply:
With `assert np.all(test.to_array()==np.array(_list_dense))` it fails for:

FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[ket-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[bra-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[square-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[wide-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[tall-float64] - assert False


Could this be due to the casting rules? I looks weird as it only fails for float64.